### PR TITLE
Added reflection-free API for weak events

### PR DESCRIPTION
**Pull request by kekekeks**
## What does the pull request do?

Adds special WeakEvent type that's constructed from the usual event pattern and manages weak subscriptions for that particular event

## What is the current behavior?
WeakSubscriptionManager and WeakEventHandlerManager both rely on reflection, which is not really compatible with trimming and performance

### CODE REVIEW

**Comment by kekekeks** [comment]
Visual and Pen will be updated in a separate PR.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0017702-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

File: src/Avalonia.Base/Utilities/WeakEvent.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                ScheduleCompact();
+            }
+        }
+
+        void ScheduleCompact()
+        {
+            if(_compactScheduled)
+                return;
+            _compactScheduled = true;
+            Dispatcher.UIThread.Post(Compact, DispatcherPriority.Background);
```
Comment (MarchingCube) [inline src/Avalonia.Base/Utilities/WeakEvent.cs:113]: Might be worth to cache a delegate to `Compact`.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0017725-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**